contracts/rust/adapter/src/sol_types.rs

```diff
@@ -0,0 +1,207 @@
+//! Solidity types for interacting with contracts
+//! Re-export types that are used, sometimes renamed to avoid collision.
+//!
+//! NOTE: Due to <https://github.com/foundry-rs/foundry/issues/10153>,
+//! try to re-export the same type from the "youngest" child contract since that is the contract whose functions are being called,
+//! thus from whom the rust bindings are expected.
+//! E.g. Both PlonkVerifier and LightClient, and LightClientV2 depends on BN254. The inheritance relationship is:
+//!   BN254 <- PlonkVerifier <- LIghtClient <- LightClientV2
+//! Most of the time, we interact with PlonkVerifier's function via LightClientV2, thus import BN254.G1Point from `bindings::plonkverifierv2`.
+//! When we need to directly interact with PlonkVerifier's method, implement stupid plain `From<lc2::BN254::G1Point> for pv::BN254::G1Point`.
+//! If you are lazy, you can even use unsafe memory transmute since they are literally the same representation, duplicated in different modules,
+//! thus treated by the rust type systems as distinct types.
+//!
+//! Another usage is in the differential testing in Solidity tests. In those cases, the actual types don't matter, since they will all `abi_encode()`
+//! into the exact same bytes before being communicated over to contract via FFI. Thus using any one of them is fine.
+
+use alloy::sol;
+
+/// # What to re-export, what to hide?
+/// - export contract struct itself, but try to avoid export instance type (instead, use ::new() to get a handle)
+/// - avoid exporting `xxCall` and `xxReturn` types, they usually can be converted/transmuted from existing struct
```

Comment:
It would be better to say when types can be safely transmuted. If we didn't want to use transmute for this, how much extra code would it be?

Reply:
it's all just a matter of placating the type system, type conversion is needed. Of course everything can be down without `transmute`, but the code will be extremely verbose. you have to map field-by-field between 4 different types of `BN254::G1Point`.